src/Specular/Internal/Incremental/Node.purs

```diff
@@ -0,0 +1,148 @@
+--     cog.outl("""
+--     foreign import get_%(name)s :: forall a. EffectFn1 (Node a) (%(ty)s)
+--     foreign import set_%(name)s :: forall a. EffectFn2 (Node a) (%(ty)s) Unit
+--     """ % {'name': name, 'ty': ty})
+-- ]]]
+
+foreign import get_dependents :: forall a. EffectFn1 (Node a) (MutableArray SomeNode)
+
+
+foreign import get_observers :: forall a. EffectFn1 (Node a) (MutableArray (Observer a))
+
+
+foreign import get_source :: forall a. EffectFn1 (Node a) (Source a)
+
+
+foreign import get_adjustedHeight :: forall a. EffectFn1 (Node a) (Int)
+foreign import set_adjustedHeight :: forall a. EffectFn2 (Node a) (Int) Unit
+
+
+foreign import get_changedAt :: forall a. EffectFn1 (Node a) (Int)
+foreign import set_changedAt :: forall a. EffectFn2 (Node a) (Int) Unit
+
+
+foreign import get_height :: forall a. EffectFn1 (Node a) (Int)
+foreign import set_height :: forall a. EffectFn2 (Node a) (Int) Unit
+
+
+foreign import get_name :: forall a. EffectFn1 (Node a) (String)
+foreign import set_name :: forall a. EffectFn2 (Node a) (String) Unit
+
+
+foreign import get_value :: forall a. EffectFn1 (Node a) (Optional a)
+foreign import set_value :: forall a. EffectFn2 (Node a) (Optional a) Unit
+
+-- [[[end]]]
+
+foreign import _new ::
+  forall a.
+  EffectFn6
+    (Optional Any) -- Optional.none
+    (Source a)
+    (MutableArray SomeNode)
+    (MutableArray (Observer a))
+    (Optional a)
+    Int
+  (Node a)
+
+-- * Node source
+
+type Source a =
+  { compute :: EffectFn1 (Node a) (Optional a)
+    -- Compute the node value, if none is returned then we don't change value and don't propagate.
+  , dependencies :: Effect (Array SomeNode)
+  }
+
+-- * Existential
+
+type SomeNode = Node Any
+
+toSomeNode :: forall a. Node a -> SomeNode
+toSomeNode = unsafeCoerce
+
+toSomeNodeArray :: forall a. Array (Node a) -> Array SomeNode
+toSomeNodeArray = unsafeCoerce
+
+-- * Creation
+
+create :: forall a. EffectFn1 (Source a) (Node a)
+create = mkEffectFn1 \source -> do
+  dependents <- MutableArray.empty
+  observers <- MutableArray.empty
+  runEffectFn6 _new
+    Optional.none
+    source
+    dependents
+    observers
+    Optional.none -- value
+    0             -- height
+
+-- * Utils
+
+refcount :: forall a. EffectFn1 (Node a) Int
+refcount = mkEffectFn1 \node -> do
+  observers <- runEffectFn1 get_observers node
+  numDependents <- runEffectFn1 MutableArray.length observers
+  dependents <- runEffectFn1 get_dependents node
+  numObservers <- runEffectFn1 MutableArray.length dependents
+  pure (numDependents + numObservers)
+
+valueExc :: forall a. EffectFn1 (Node a) a
+valueExc = mkEffectFn1 \node -> do
+  value_opt <- runEffectFn1 get_value node
+  pure (Optional.fromSome value_opt)
+
+annotate :: forall a. EffectFn2 (Node a) String Unit
+annotate = set_name
+
+name :: forall a. Node a -> String
+name node = unsafePerformEffect (runEffectFn1 get_name node)
+
+isChangingInCurrentStabilization :: forall a. EffectFn1 (Node a) Boolean
```

Comment:
Always `-1`?

Reply:
No, it is not -1 during stabilization; see above.